rust_src/src/marker.rs

```diff
@@ -341,12 +337,121 @@ pub extern "C" fn marker_byte_position(marker: Lisp_Object) -> ptrdiff_t {
 /// whether BUFFER is a buffer object and return buffer pointer
 /// corresponding to BUFFER if BUFFER is live, or NULL otherwise.
 #[no_mangle]
-pub extern "C" fn live_buffer(buffer: Lisp_Object) -> *mut Lisp_Buffer {
-    let mut b = LispObject::from_raw(buffer).as_buffer_or_current_buffer();
+pub extern "C" fn live_buffer(buffer: Lisp_Object) -> Option<LispBufferRef> {
+    let b = LispObject::from_raw(buffer).as_buffer_or_current_buffer();
     if b.is_live() {
-        b.as_mut()
+        Some(b)
+    } else {
+        None
+    }
+}
+
+/// Internal function to set MARKER in BUFFER at POSITION.  Non-zero
+/// RESTRICTED means limit the POSITION by the visible part of BUFFER.
+fn set_marker_internal(
+    marker: LispObject,
+    position: LispObject,
+    buffer: LispObject,
+    restricted: bool,
+) -> LispObject {
+    let buf = live_buffer(buffer.to_raw());
+    let mut m = marker.as_marker_or_error();
+
+    // Set MARKER to point nowhere if BUFFER is dead, or
+    // POSITION is nil or a marker points to nowhere.
+    if position.is_nil()
+        || position
+            .as_marker()
+            .map_or(false, |m| !m.buffer().is_some()) || !buf.is_some()
+    {
```

Comment:
`!is_some` is better written as `is_none`.

Reply:
Twice this code potentially converts to marker and then extracts the buffer. Would this not read better if that buffer were stored in a temporary?

Reply:
Good point about using a more specific type in the argument list.

I missed that marker and position could be in different buffers. Thank you.

Reply:
> A minor improvement would be to declare marker as a LispMarkerRef in the argument list, to convert automatically. Similarly, position could be an Option<LispMarkerRef>.

I thought this only works with lisp functions ?

Reply:
automagically yes. But you can do it explicitly.

Reply:
The point of a type system is to have strong fences.

Reply:
> A minor improvement would be to declare marker as a LispMarkerRef in the argument list, to convert automatically. 

If I pass a `LispMarkerRef` I have to do the type checking in the functions that are calling `set_marker_internal` ? I don't know how to convert it automatically here.

> Similarly, position could be an Option<LispMarkerRef>.

`position` can be an nil, a marker or an `EmacsInt`. I'm not sure how to do this with an ` Option<LispMarkerRef>`.